File: csharp-version/DesafioR6Geral/WinMatriculaApp/MatriculaTeste.cs
using System;
using System.Collections.Generic;
using System.Globalization;
using System.IO;
using System.Linq;
using System.Text;
using System.Text.RegularExpressions;
using System.Threading.Tasks;

namespace WinMatriculaApp
{
    public class MatriculaTeste
    {
        //Método para verificar somente uma matrícula.
        public static string InformarMatricula(string matricula)
        {
            return MatriculaTeste.VerificarMatricula(matricula);
        }

        //Método para verificar matrícula dentro de um arquivo.
        public static int VerificarMatriculaArquivo(string path)
        {
            try
            {
                if (File.Exists(path))
                {
                    List<string> list = new List<string>();

                    foreach (string matricula in File.ReadAllLines(path))
                    {
                        list.Add(MatriculaTeste.VerificarMatricula(matricula));
                    }

                    EnviarParaArquivo(1, list);
                    return 1; // Retorno OK;
                }
                else
                {
                    return 0; // Nenhum arquivo informado;
                }
            }
            catch (Exception)
            {
                return -1; // Erro gerado;
            }
        }

        //Método para testar se as verificações das matrícula num arquivo estão corretas.
        public static int TestarVerificacaoMatriculaArquivo(string path)
        {
            try
            {
                if (File.Exists(path))
                {
                    List<string> list = new List<string>();
                    String result = "";

                    foreach (string matricula in File.ReadAllLines(path))
                    {
                        if((matricula == MatriculaTeste.VerificarMatricula(matricula.Substring(0, 4))))
                        {
                            result = "Verdadeiro";
                        }
                        else
                        {
                            result = "Falso";
                        }
                        list.Add(matricula + " - " + result);
                    }

                    return EnviarParaArquivo(2, list);
                }
                else
                {
                    return 0; //Arquivo não encontrado;
                }
            }
            catch (Exception)
            {
                return -1;   //Erro gerado; 
            }
        }

        //Método utilizado para salvar as informãções em arquivo.
        private static int EnviarParaArquivo(int opc, List<string> matriculas)
        {
            try
            {
                StreamWriter file = null;

                //Criando arquivo de saída de verificação de matrícula.
                if (opc == 1)
                {
                    file = new StreamWriter(@"matriculasComDV.txt");
                    foreach (string matricula in matriculas)
                    {
                        file.WriteLine(matricula);
                    }
                }

                //Criando arquivo de saída de teste da verificação de matrícula.
                if (opc == 2)
                {
                    file = new StreamWriter(@"matriculasVerificadas.txt");
                    foreach (string matricula in matriculas)
                    {
                        file.WriteLine(matricula);
                    }
                }

                file.Close();

                return 1;
            }
            catch (Exception)
            {
                return -1;
            }
        }

        //Método para validar entrada de uma matrícula.
        private static bool ValidarMatricula(String matricula)
        {
            string pattern = "^[0-9]{4}$";
            Regex rgx = new Regex(pattern, RegexOptions.None);
            return rgx.IsMatch(matricula);
        }

        //Método que faz a verificação da matrícula.
        private static string VerificarMatricula(String matricula)
        {
            try
            {
                if(MatriculaTeste.ValidarMatricula(matricula))
                {
                    int i = 6;
                    int soma = 0;

                    foreach (char item in matricula)
                    {
                        soma = soma + Int32.Parse(item.ToString()) * (--i);
                    }

                    int div = soma % 16;
                    string hexValue = div.ToString("X");
                    return String.Format("{0}-{1}", matricula, hexValue);
                }
                else
                {
                    return String.Format("{0} - MATRICULA INVÁLIDA.", matricula);
                }
            }
            catch (Exception)
            {
                throw;
            }
        }
    }
}

File: java-version/JavaMatriculaApp/src/main/java/com/javamatriculaapp/MatriculaTeste.java
package com.javamatriculaapp;

import java.io.BufferedReader;
import java.io.BufferedWriter;
import java.io.File;
import java.io.FileNotFoundException;
import java.io.FileReader;
import java.io.FileWriter;
import java.io.IOException;
import java.util.ArrayList;
import java.util.List;
import java.util.logging.Level;
import java.util.logging.Logger;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

public class MatriculaTeste {

    //Método para calcular o dígito verificador de somente uma matrícula.
    public static String InformarMatricula(String matricula) {
        return MatriculaTeste.CalcularMatricula(matricula);
    }

    //Método para calcular o dígito verificador de matrículas dentro de um arquivo.
    public static int CalcularMatriculaArquivo(String path) {
        try {
            int i = 0;
            List<String> matriculas = MatriculaTeste.ObterMatriculasArquivo(path);
            if (matriculas != null) {
                for (String matricula : matriculas) {
                    matriculas.set(i, CalcularMatricula(matricula));
                    i = i + 1;
                }
            }
            return GravarEmArquivo(1, matriculas); // 1 OK or -1 Error
        } catch (Exception e) {
            return -1; // Erro gerado;
        }
    }

    //Método para testar se as verificações das matrícula no arquivo estão corretas.
    public static int TestarVerificacaoMatriculaArquivo(String path) {
        try {
            int i = 0;
            List<String> matriculas = MatriculaTeste.ObterMatriculasArquivo(path);
            if (matriculas != null) {
                String result = "";
                for (String matricula : matriculas) {
                    if(matricula.equals(CalcularMatricula(matricula.substring(0, 4))))
                    {
                        result = "Verdadeiro";
                    }
                    else
                    {
                        result = "Falso";
                    }
                    matriculas.set(i, String.format("%s - %s", matricula, result));
                    i = i + 1;
                }
            }
            return GravarEmArquivo(2, matriculas); // 1 OK or -1 Error
        } catch (Exception e) {
            return -1; // Erro gerado;
        }
    }

    //Método para obter as matriculas listadas num arquivo.
    private static List<String> ObterMatriculasArquivo(String path) {

        File file = new File(path);

        if (file.exists()) {
            List<String> list = new ArrayList<>();
            FileReader fileReader = null;
            try {
                fileReader = new FileReader(file);
            } catch (FileNotFoundException ex) {
                Logger.getLogger(MatriculaTeste.class.getName()).log(Level.SEVERE, null, ex);
            }
            BufferedReader bufferedReader = new BufferedReader(fileReader);
            String line;
            try {
                while ((line = bufferedReader.readLine()) != null) {
                    list.add(line);
                }
            } catch (IOException ex) {
                Logger.getLogger(MatriculaTeste.class.getName()).log(Level.SEVERE, null, ex);
            }
            return list;
        } else {
            return null;
        }
    }

    //Método que faz o calculo do dígito verificador da matrícula.
    private static String CalcularMatricula(String matricula) {
        try {
            int i = 6;
            int soma = 0;

            for (char item : matricula.toCharArray()) {
                soma = soma + (Integer.parseInt(String.valueOf(item)) * (--i));
            }
            int div = soma % 16;
            String hex = Integer.toHexString(div);
            return String.format("%s-%s", matricula, hex.toUpperCase());
        } catch (NumberFormatException e) {
            return "";
        }
    }

    //Método para validar entrada de uma matrícula.
    public static Boolean ValidarMatricula(String matricula) {
        String pattern = "^[0-9]{4}$";
        Pattern r = Pattern.compile(pattern);
        Matcher m = r.matcher(matricula);
        return m.matches();
    }

    //Método para gravar lista de matrícula em arquivo.
    private static int GravarEmArquivo(int opc, List<String> matriculas) {
        String name = "";
        if (opc == 1) {
            name = "matriculasComDV.txt";
        }

        if (opc == 2) {
            name = "matriculasVerificadas.txt";
        }

        try {
            try (BufferedWriter writer = new BufferedWriter(new FileWriter(new File(name)))) {
                for (String matricula : matriculas) {
                    writer.append(matricula);
                    writer.newLine();
                }
            }
            return 1;
        } catch (IOException ex) {
            Logger.getLogger(MatriculaTeste.class.getName()).log(Level.SEVERE, null, ex);
            return -1;
        }
    }
}
